api_tutorials/bait_cnn/bait_performance_cnn.py

```diff
@@ -0,0 +1,605 @@
+
+    model.add(Dropout(0.3))
+    model.add(Conv1D(filters=64, kernel_size=16, activation="relu", init='normal', padding='valid'))
+    model.add(Dropout(0.3))
+    model.add(MaxPooling1D(pool_size=3, strides=3))
+    model.add(Flatten())
+
+    model.add(Dense(units=32, kernel_initializer='normal'))
+    model.add(Activation('relu'))
+
+    model.add(Dense(units=1, kernel_initializer=RandomNormal(mean=1.0, stddev=0.5, seed=None)))
+    model.add(Activation("relu"))
+
+    adamo = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, clipnorm=1.)
+    my_metrics = [metrics.mean_squared_error, rmse_log]
+
+    model.compile(loss='mean_squared_error', optimizer=adamo, metrics=my_metrics)
+    model.summary()
+
+    return model
+
+
+def build_small_functional_bait_model(args):
+    bait_shape = (args.window_size, len(args.inputs),)
+    annotation_shape = (len(args.annotations),)
+
+    print("Bait shape: %s" % (bait_shape,))
+    input_baits = Input(name="bait", shape=bait_shape)
+
+    x = Conv1D(activation='relu',
+               padding='valid',
+               filters=100,
+               kernel_size=8,
+               kernel_initializer='normal', name='conv1')(input_baits)
+
+    x = Dropout(0.3)(x)
+
+    x = Conv1D(kernel_initializer="normal", activation="relu", padding="valid", filters=64, kernel_size=8)(x)
+    x = Dropout(0.3)(x)
+    x = MaxPooling1D(pool_size=3, strides=3)(x)
+
+    x = Conv1D(kernel_initializer="normal", activation="relu", padding="valid", filters=40, kernel_size=8)(x)
+    x = Dropout(0.3)(x)
+    x = MaxPooling1D(pool_size=3, strides=3)(x)
+
+    x = Flatten()(x)
+
+    x = Dense(units=32, kernel_initializer="normal", activation="relu")(x)
+
+    print("Annotation shape: %s" % (annotation_shape,))
+
+    input_annotations = Input(name="annotation", shape=annotation_shape)
+
+    xy = Concatenate(axis=-1)([x, input_annotations])
+
+    xy = Dense(units=32, kernel_initializer="normal", activation="relu")(xy)
+
+    predictions = Dense(units=1, kernel_initializer=RandomNormal(mean=1.0, stddev=0.5, seed=None), activation="relu")(
+        xy)
+    my_metrics = [metrics.mean_squared_error, rmse_log, gme]
+
+    # this creates a model that includes
+    # the Input layer and three Dense layers
+    model = Model(inputs=[input_baits, input_annotations], outputs=predictions)
+
+    # # add some TensorBoard annotations
+    # conv1d_1 = filter(lambda y: y.name == "conv1d_1", model.layers)[0]
+    # conv1d_1_shape = map(lambda x: x.value, conv1d_1.kernel.get_shape())
+    # conv1d_1_shape.insert(0, 1)
+    #
+    # reshaped = tf.reshape(conv1d_1.kernel, conv1d_1_shape)
+    #
+    # filters=put_kernels_on_grid(reshaped, 2)
+    #
+    # merged = tf.summary.merge_all()
+    # train_writer = tf.summary.FileWriter("./log/" + '/train')
+    #
+    adamo = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, clipnorm=1.)
+    model.compile(loss=metrics.mean_squared_error, optimizer=adamo, metrics=my_metrics)
```

Comment:
I'm wondering if we should try this as a classification problem by binning the coverage first and then trying to predict (high, med, low) or (too high, high, medium, low, too low) or something.  Then we could use categorical crossentropy as our loss, which tends to have more well-behaved learning dynamics, also we could use categorical accuracy as a metric to get a quick idea of how the model compares to chance...

Reply:
and then what?  modify the loss so that it tries to maximize xentropy + MSE?

Reply:
I was thinking to just try categorical crossentropy though we could do it a multi-task problem where the model tries to do both.  It is easy to do with the functional API. Something like:
```regression = Dense(units=1, kernel_initializer=RandomNormal(mean=1.0, stddev=0.5, seed=None), activation="relu")(xy)
classification = Dense(units=5, activation='softmax')(xy)
model = Model(inputs=[input_baits, input_annotations], outputs=[regression, classification])
model.compile(loss=['categorical_crossentropy', metrics.mean_squared_error], optimizer=adamo, metrics=my_metrics)```